### PR TITLE
/collection_versions/all/ endpoint is now streamed

### PR DESCRIPTION
**Pull request by gerrod3**
This is super fast I think. Potentially fast enough to rethink when or if we will need publications.

### CODE REVIEW

**Review by rochacbruno — APPROVED** [review]

**Review by daviddavis — APPROVED** [review]

**Comment by daviddavis** [comment]
Is it correct that the /collection_versions/all/ no longer accepts a page option? If so, we probably need a removal changelog entry.

**Comment by gerrod3** [comment]
> Is it correct that the /collection_versions/all/ no longer accepts a page option? If so, we probably need a removal changelog entry.

Pretty sure it never accepted a page option. This endpoint is suppose to be unpaginated, but the code still had remnants of the paginated code, probably from copying and pasting.

**Comment by gerrod3** [comment]
@rochacbruno Still doing some testing to see if I can get it even faster.  With this change I no longer get timed-out when accessing /collection_versions/all/ with 3000 collection_versions. Going to test it with 6000 collection versions and check that it still works, but I'm feeling that this might be the best current solution we can go with. Here are some results using django-silk and a distribution with 94 collection versions.

**Current**
Screenshot from 2021-04-15 17-46-17
**Streamed**
Screenshot from 2021-04-16 09-00-18

One other thing is that the streamed response doesn't use DRF's fancy styling, it looks pretty bad in the browser. Maybe I could fix this, but I don't think it's such a big deal.

File: pulp_ansible/app/galaxy/v3/views.py [hunk 1]
```diff
@@ -6,10 +6,12 @@
 from django.db.models import F
 from django.db.models.expressions import Window
 from django.db.models.functions.window import FirstValue
+from django.http import StreamingHttpResponse
```
Comment (fao89) [inline pulp_ansible/app/galaxy/v3/views.py:9]: cool!

File: pulp_ansible/app/galaxy/v3/views.py [hunk 2]
```diff
@@ -446,15 +448,12 @@ def list(self, request, *args, **kwargs):
...
-            serializer = self.get_serializer(page, many=True, context=context)
-            return self.get_paginated_response(serializer.data)
-
-        serializer = self.get_serializer(queryset, many=True, context=context)
-        return Response(serializer.data)
+        cvs_template_string = "[{{ versions|map('tojson')|join(',') }}]"
+        cvs_template = Template(cvs_template_string)
+        serialized_map = map(lambda x: self.get_serializer(x, context=context).data, queryset)
+        streamed = map(lambda x: x.encode("utf-8"), cvs_template.stream(versions=serialized_map))
+        return StreamingHttpResponse(streamed)
```
Comment (fao89) [inline pulp_ansible/app/galaxy/v3/views.py:456]: should we take these to `UnpaginatedCollectionViewSet` also? Or the timeout only happens here?

**Review by fao89 — COMMENTED** [review]
This is looking pretty good, why does it still as draft?

**Comment by gerrod3** [comment]
@fao89 Should I tag this PR with the original issue for the unpaginated endpoint timeout https://pulp.plan.io/issues/8439?

**Comment by fao89** [comment]
> @fao89 Should I tag this PR with the original issue for the unpaginated endpoint timeout https://pulp.plan.io/issues/8439?

I think so!

**Review by fao89 — APPROVED** [review]
Thank you!

**Review by bmbouter — APPROVED** [review]
This is really nice! Great job!

I believe we still need publications for 2 reasons:
* to resolve "the deprecated field issues"
* pre-computing something this large once and saving it is more efficient and will speed up the sync too. This is more of a nice-to-have really.

**Pull request merged**